child_compassion/models/project_reservation.py

```diff
@@ -37,30 +37,53 @@ class ProjectReservation(models.Model):
     @api.model
     def create(self, vals):
         res = super(ProjectReservation, self).create(vals)
-        message_obj = self.env['gmc.message.pool']
-        action_id = self.env.ref('child_compassion.create_reservation').id
+        self.handle_reservation('child_compassion.create_reservation', res)
+        return res
 
-        message_vals = {
-            'action_id': action_id,
-            'object_id': res.id,
-        }
-        message_obj.with_context(async_mode=False).create(message_vals)
+    @api.multi
+    def write(self, vals):
+        res = super(ProjectReservation, self).write(vals)
+        if not self.env.context.get('creating') and self.active:
+            self.with_context(creating=False).handle_reservation(
+                'child_compassion.create_reservation', res)
         return res
 
     @api.multi
     def unlink(self):
-        message_obj = self.env['gmc.message.pool']
-        action_id = self.env.ref('child_compassion.cancel_reservation').id
+        self.cancel_reservation()
+        return
 
+    @api.multi
+    def cancel_reservation(self):
         self.active = False
-        message_vals = {
-            'action_id': action_id,
-            'object_id': self.id
+        self.handle_reservation('child_compassion.cancel_reservation')
+        return {
+            'type': 'ir.actions.act_window',
+            'view_type': 'form',
+            'view_mode': 'tree, form',
+            'res_model': 'icp.reservation',
+            'target': 'current',
         }
-        message_obj.with_context(async_mode=False).create(message_vals)
-        return
 
     @api.model
     def process_commkit(self, commkit_data):
         # TODO Implement
         return False
+
+    ##########################################################################
+    #                             PUBLIC METHODS                             #
+    ##########################################################################
+    def handle_reservation(self, action, res=False):
+        message_obj = self.env['gmc.message.pool']
+
+        action_id = self.env.ref(action).id
+        object_id = res.id if type(res) is not bool else self.id
+        message_vals = {
+            'action_id': action_id,
+            'object_id': object_id
+        }
+        pool = message_obj.with_context(async_mode=False,
+                                        creating=True).create(message_vals)
+        if pool.failure_reason:
+            raise Warning(_("Reservation impossible"), _(pool.failure_reason))
```

Comment:
You don't need to translate (`_()`) the failure message, it won't likely be translatable...